pilot/pkg/proxy/envoy/gateway.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"strings"
+
+	"reflect"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	routing "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+func buildGatewayListeners(mesh *meshconfig.MeshConfig, instances []*model.ServiceInstance, config model.IstioConfigStore, ingress model.Node) Listeners {
+	gateways, err := config.List(model.Gateway.Type, model.NamespaceAll)
+	if err != nil {
+		log.Warnf("error listing gateways: %v", err)
+		return Listeners{}
+	} else if len(gateways) == 0 {
+		log.Debugf("listing gateways in namespace %q, got none", model.NamespaceAll)
+		return Listeners{}
+	}
+
+	gateway := gateways[0].Spec.(*routing.Gateway)
+	for _, spec := range gateways[1:] {
+		model.MergeGateways(gateway, spec.Spec.(*routing.Gateway))
+	}
```

Comment:
I think this is trouble. In routing rules, we can refer to gateways by name. In other words, the routing rules can pick source pods or the gateways where they have to be applied. This merge thingy here is going to eliminate gateway objects.

One way to approach this would be to convert the gateway into our service model, wherein the gateway name becomes a non addressable service name, and the workload or instance where gateway is running has a set of internally created labels that corresponds to the set of gateways running on that workload. Once you have this you could essentially reuse the rest of routing logic, that matches a rule.source to a pod/vm/cf app based on the instance's labels.

Reply:
So my reasoning here is this:
- We know multiple Gateways don't define the same/overlapping hostnames due to our validation. Each host can only be claimed by a single Gateway.
- LDSv2 which has the notion of filter chain per host (kinda; as I understand it, it will after Piotr's PRs relating to SNI sniffing land?)
- RouteRules apply to Gateways by name only in Istio config itself
- Therefore, we can wire up all Gateways into a single listener that dispatches via host
  - routes can be configured per host; we just keep track of all the Gateway names in our composite for when we build the routes

I hadn't finished implementing the second half, keeping track of Gateway names and using them for route construction. I also hand't done any of the LDSv2 things, which will be required eventually (AFAICT for Ingress case we can get by since the filter chain is duplicated for all Ingresses since they're HTTP/HTTPS only).

Reply:
(That's why I hadn't submitted the WIP PR for comments yet - that initial commit really was just the first pass through for some of the basic logic.)

Reply:
> We know multiple Gateways don't define the same/overlapping hostnames due to our validation. Each host can only be claimed by a single Gateway

Zack, I don't think that is right. Unlike other things in routing, Gateway has an implicit concept of physical network affiliation. I could have an external/public and internal (kube-monolith-vm) gateway exposing same hostname. Assuming that the DNS infrastructure resolves to the correct gateway (e.g., requests for foo.bar.com inside the mesh get resolved to internal gateway), users now have the nice property of being able to control gateway's network visibility based on whatever access restrictions people have in their networks.

Essentially, if we treat the gateway as a pseudo service with name == gatewayname, but hostnames == gateway.hostnames, then we can play around with the network locality of the gateway (egress/ingress/middle/etc.).